src/libraries/System.Private.CoreLib/src/System/IO/BinaryWriter.cs

```diff
@@ -58,10 +56,9 @@ public BinaryWriter(Stream output, Encoding encoding, bool leaveOpen)
                 throw new ArgumentException(SR.Argument_StreamNotWritable);
 
             OutStream = output;
-            _buffer = new byte[16];
             _encoding = encoding;
-            _encoder = _encoding.GetEncoder();
             _leaveOpen = leaveOpen;
+            _useFastUtf8 = encoding.IsUTF8CodePage && encoding.EncoderFallback.MaxCharCount <= 1;
```

Comment:
Would it be safer to just do a reference comparison against Encoding.UTF8?  Seems like that would cover the 99% case while not being susceptible to problems with, for example, someone creating a custom encoding that's UTF8 but that does something special in its encoding methods, like logging?

Reply:
I don't want to compare against only `Encoding.UTF8` because it doesn't hit the case where somebody passed `new UTF8Encoding(/* ... */)`. Both of these patterns are very common in .NET code. I feel that if we special-case the former we should also special-case the latter.

In reality, what I care about is that: (a) the implementation says it's UTF-8; and (b) if an error is encountered, the implementation will either throw or emit no more than a single character like `'?'` or `'\uFFFD'`. If both of these are true, then I know the worst-case possible expansion of the input data is 1 char -\> 3 bytes. (If the implementation claims to be UTF-8 but actually does something else to violate the above assumption, bad things happen.)

In an ideal world I could call `Encoding.GetMaxByteCount(1)`, compare it against 3 to set the "fast-path" flag, and call it a day. But `Encoding.UTF8.GetMaxByteCount(1)` weirdly returns 6 instead of 3 for historical reasons. A more straightforward way to check this would be to say `if (e is UTF8Encoding /* matches subclasses as well */ && e.EncoderFallback.MaxCharCount <= 1)`, but per an earlier suggestion from @tarekgh a few months back it seems like querying the virtual `CodePage` property may be more reliable than checking the type hierarchy.